### PR TITLE
Throw ArgumentNullExceptions for null arguments

### PR DESCRIPTION
**Pull request by stephentoub**
### Motivation and Context

Use ArgumentNullException as devs expect, and avoid exception-related work when not throwing.

### Description

.NET developers expect that passing null erroneously to a method will result in an ArgumentNullException, but currently it's resulting in a custom ValidationException.

There are also places where string interpolation is being used to create the error message; the work and allocation associated with that interpolation is going to happen regardless of whether the exception is thrown or not.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Comment by MovGP0** [comment]
LGTM. 

💡 A code weaver like [Fody.NullGuard](https://github.com/Fody/NullGuard) or a [Roslyn code generator](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md) might be helpful in replacing all those runtime `null` checks.

File: dotnet/src/SemanticKernel.Abstractions/Diagnostics/Verify.cs [hunk 1]
```diff
@@ -15,18 +15,18 @@ internal static class Verify
     private static readonly Regex s_asciiLettersDigitsUnderscoresRegex = new("^[0-9A-Za-z_]*$");
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply dotnet/src/SemanticKernel.Abstractions/Diagnostics/Verify.cs:17]: > Thoughts on having a ValidateNotNull for the previous behavior?

We could... but most of the places that were using it for things other than actual arguments were doing things like this:
```C#
                config = PromptTemplateConfig.FromJson(File.ReadAllText(configPath));
                Verify.NotNull(config, $"Invalid prompt template configuration, unable to parse {configPath}");
```
That means it's going to be allocating a whole new string for the message even in the success case where config is non-null.  By providing a helper like this, I worry it'll encourage that kind of misuse.

[the remark this answers is not in the document]
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel.Abstractions/Diagnostics/Verify.cs:17]: Sounds good to me, thanks Stephen!

File: dotnet/src/SemanticKernel/Planning/FunctionFlowRunner.cs [hunk 2]
```diff
@@ -259,8 +259,10 @@ private async Task<string> ProcessNodeListAsync(XmlNodeList nodeList, string fun
 
                     if (processFunctions && !string.IsNullOrEmpty(functionName))
                     {
-                        Verify.NotNull(functionName, nameof(functionName));
-                        Verify.NotNull(skillFunction, nameof(skillFunction));
+                        if (skillFunction is null)
```
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/Planning/FunctionFlowRunner.cs:262]: Someone will need to resolve conflict with https://github.com/microsoft/semantic-kernel/pull/523 -- making note for self.

**Review by lemillermicrosoft — DISMISSED** [review]

File: dotnet/src/SemanticKernel/Planning/FunctionFlowRunner.cs (hunk 2 above)
Comment (lemillermicrosoft) [reply dotnet/src/SemanticKernel/Planning/FunctionFlowRunner.cs:262]: I beat you :)

See `Plan.cs`, `SequentialPlanParser.cs`, and `SequentialPlanner.cs` for applying changes.

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public abstract class ClientBase
         CompleteRequestSettings requestSettings,
         CancellationToken cancellationToken = default)
     {
-        Verify.NotNull(requestSettings, "Completion settings cannot be empty");
+        Verify.NotNull(requestSettings);
```
Comment (dluc) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:43]: Wouldn't you agree that custom error messages would help developers understanding the context, e.g. as opposed to generic exceptions and line numbers?

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs:43]: My changes include using CallerArgumentExpression with NotNull helper. The exception message in this case will be:
```
Value cannot be null. (Parameter 'requestSettings')
```
I think that's more valuable, telling the developer exactly which parameter is the problem. With the custom message:
```
Completion settings cannot be empty
```
the developer needs to know that a) empty means null and b) "completion settings" is referring to the requestSettings parameter.

**Review by lemillermicrosoft — DISMISSED** [review]

**Comment by stephentoub** [comment]
> looks like there are some conflicts that need to be resolved before merging.

Done. Thanks.

**Review by shawncal — APPROVED** [review]

**Pull request merged**